### ISSUE TITLE
Generate API Token and generate new variable in playbook to use in further tasks

### ISSUE THREAD

**Issue by carlos-segundo** [fortinet-ansible-dev/ansible-galaxy-fortios-collection#90]
Hi,

I'll have to do full automation of fortigate hw appliances including bootstrap and provisioning and full lifecycle.

Now using the new galaxies many API calls can be used via the modules (_**is there a list of which API calls are not yet implemented and if there is work to support them ongoing?**_).

Since Fortinet recommends to use the API Token for the Fortigate it'll complicate the task and maybe we'd rather use the username and password method to continue (_**except if it gets removed with the next major release?**_).

1) Hardware Boots (Bootmenu -> can be automated using console connection)
2) Default IP and default admin user -> Can the first login where one is now forced to change the password be done via Ansible or is this something that is a must to be done via console/ssh?
3) Create new accprofile for rest api (Optional) -> via Ansible accprofile using user+password
4) Create api_user -> via Ansible api_user using user+password

5) **Generate API Key**

Now I have not found a module which would allow this using the rest api call "/api/v2/monitor/system/api-user/generate-key?vdom=root"

Currently I would have to use uri module? The return value should be stored in an updated hosts file, so that from now on the tasks are run with the API Token and not anymore with the user + password. How to best approach this?

6) Change default admin password, (or delete and create a new one) using API Token
7) All other changes are then done using the API Token

Are others using it this way or just skip it and continue to use username and password?
Did I miss a module that allows to get the API Key?

PS: It's interesting that one of the biggest benefits described to use API Token, which was the Trusted hosts requirement is gone in 6.4.x, i.e. I can just create an api user and use it with no Trusted Host specification.

**Comment by mbdraks** [comment]
@carlos-segundo 

Here's how you can create the api-user, generate the token and update the relevant host_vars:

```yaml
---
- name: Login with credentials | Legacy login (FOS <= 6.4.2)
  ansible.builtin.uri:
    url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/logincheck
    validate_certs: false
    method: POST
    body: "username={{ ansible_user }}&secretkey={{ ansible_password }}"
  register: login

- name: Configure API users with credentials
  fortinet.fortios.fortios_system_api_user:
    state: present
    system_api_user:
      name: "{{ api_user_name }}"
      accprofile: "{{ accprofile }}"
      trusthost:
        - id: "{{ trusthost_id }}"
          ipv4_trusthost: "{{ ipv4_trusthost }}"
      vdom:
        - name: "{{ vdom }}"

- name: Generate API Key
  ansible.builtin.uri:
    url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/api/v2/monitor/system/api-user/generate-key
    validate_certs: false
    method: POST
    body_format: json
    body: {"api-user": "{{ api_user_name }}"}
    headers:
      # Strip double-quote from csrftoken string
      X-CSRFTOKEN: "{{ login.cookies.ccsrftoken | regex_replace('\"', '')  }}"
      Cookie: "{{ login.cookies_string }}"
  register: r

- name: Update token at host_vars file
  ansible.builtin.lineinfile:
    path: "{{ ansible_env.PWD }}/../host_vars/{{ inventory_hostname }}.yml"
    regexp: '^fortios_access_token:'
    line: "fortios_access_token: {{ r.json.results.access_token }}"
```

Here's a few details of this playbook:

1. I'm using the legacy (pre-6.4.2 new authentication api) login so it will work with any versions, it still works after 6.4.2 but will stop working on FOS 7.0. After 7.0 (or after 6.4.2 if you want to test you should use the new auth login api.
2. I'm calling the fortios_system_api_user module here without providing a token, this means that's going to use the ansible_user / ansible_password associated with this device.
3. Calling the generate-key and using the cookie created at step 1, with the traditional login
4. Updating the token on the file, here I'm replacing the line on the file so at start make sure you have at least a dummy line on the host_var like `fortios_access_token: not_valid`. Once this task runs it will update with a valid token, make sure to double check the path too, it should make sense to whatever folder structure you're using.


And here's the same playbook but using the new auth api, availble on FOS 6.4.2+, there's some additional debug commands that you can remove if you want:

```yaml
- name: Login with credentials
  ansible.builtin.uri:
    # Authentication API available on FOS 6.4.2+
    url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/api/v2/authentication
    validate_certs: false
    method: POST
    body_format: json
    body: {"username": "{{ ansible_user }}", "secretkey": "{{ ansible_password }}", "request_key": true}
  register: login

- name: Configure API users
  fortinet.fortios.fortios_system_api_user:
    state: present
    access_token: "{{ login.json.session_key }}"
    system_api_user:
      name: "{{ api_user_name }}"
      accprofile: "{{ accprofile }}"
      trusthost:
        - id: "{{ trusthost_id }}"
          ipv4_trusthost: "{{ ipv4_trusthost }}"
      vdom:
        - name: "{{ vdom }}"

- name: Generate API Key
  ansible.builtin.uri:
    url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/api/v2/monitor/system/api-user/generate-key
    validate_certs: false
    method: POST
    body_format: json
    body: {"api-user": "{{ api_user_name }}"}
    headers:
      Authorization: "Bearer {{ login.json.session_key }}"
  register: r

- name: Set new token
  set_fact:
    fortios_access_token: "{{ r.json.results.access_token }}"

- name: Display new API key
  debug:
    var: fortios_access_token

- name: Update token at host_vars file
  ansible.builtin.lineinfile:
    path: "{{ ansible_env.PWD }}/../host_vars/{{ inventory_hostname }}.yml"
    regexp: '^fortios_access_token:'
    line: "fortios_access_token: {{ fortios_access_token }}"

- name: File updated
  debug:
    msg: "File {{ ansible_env.PWD }}/../host_vars/{{ inventory_hostname }}.yml updated"
```

**Comment by mbdraks** [comment]
Another comment, for your use case I would advise the usage of FortiDeploy, this way you can have true Zero Touch Provisioning. 

The workflow would be like this:

1. Turn on the device, ISP cable connected at the wan interface of the device providing IP with DHCP
2. Device boots, call FortiCloud, identify that it should contact FMG IP: a.b.c.d
3. Device self-configure itself with FMG IP/SN and contact FMG, at this point (and considering that pre-provisioning was already done with a device model at this FMG) the device download his config and self-configure everything, including api-users/tokens etc.

This is a very short explanation, you should check with your rep about FortiCloud details if you want but reading your requirements it seems the most common use case for FDP (FortiDeploy).

**Comment by carlos-segundo** [comment]
> Another comment, for your use case I would advise the usage of FortiDeploy, this way you can have true Zero Touch Provisioning.
> 
> The workflow would be like this:
> 
> 1. Turn on the device, ISP cable connected at the wan interface of the device providing IP with DHCP
> 2. Device boots, call FortiCloud, identify that it should contact FMG IP: a.b.c.d
> 3. Device self-configure itself with FMG IP/SN and contact FMG, at this point (and considering that pre-provisioning was already done with a device model at this FMG) the device download his config and self-configure everything, including api-users/tokens etc.
> 
> This is a very short explanation, you should check with your rep about FortiCloud details if you want but reading your requirements it seems the most common use case for FDP (FortiDeploy).

In my use case we have no internet connection available, private cloud implementation.

**Comment by carlos-segundo** [comment]
> @carlos-segundo
> 
> Here's how you can create the api-user, generate the token and update the relevant host_vars:
> 
> ```yaml
> ---
> - name: Login with credentials | Legacy login (FOS <= 6.4.2)
>   ansible.builtin.uri:
>     url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/logincheck
>     validate_certs: false
>     method: POST
>     body: "username={{ ansible_user }}&secretkey={{ ansible_password }}"
>   register: login
> 
> - name: Configure API users with credentials
>   fortinet.fortios.fortios_system_api_user:
>     state: present
>     system_api_user:
>       name: "{{ api_user_name }}"
>       accprofile: "{{ accprofile }}"
>       trusthost:
>         - id: "{{ trusthost_id }}"
>           ipv4_trusthost: "{{ ipv4_trusthost }}"
>       vdom:
>         - name: "{{ vdom }}"
> 
> - name: Generate API Key
>   ansible.builtin.uri:
>     url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/api/v2/monitor/system/api-user/generate-key
>     validate_certs: false
>     method: POST
>     body_format: json
>     body: {"api-user": "{{ api_user_name }}"}
>     headers:
>       # Strip double-quote from csrftoken string
>       X-CSRFTOKEN: "{{ login.cookies.ccsrftoken | regex_replace('\"', '')  }}"
>       Cookie: "{{ login.cookies_string }}"
>   register: r
> 
> - name: Update token at host_vars file
>   ansible.builtin.lineinfile:
>     path: "{{ ansible_env.PWD }}/../host_vars/{{ inventory_hostname }}.yml"
>     regexp: '^fortios_access_token:'
>     line: "fortios_access_token: {{ r.json.results.access_token }}"
> ```
> 
> Here's a few details of this playbook:
> 
> 1. I'm using the legacy (pre-6.4.2 new authentication api) login so it will work with any versions, it still works after 6.4.2 but will stop working on FOS 7.0. After 7.0 (or after 6.4.2 if you want to test you should use the new auth login api.
> 2. I'm calling the fortios_system_api_user module here without providing a token, this means that's going to use the ansible_user / ansible_password associated with this device.
> 3. Calling the generate-key and using the cookie created at step 1, with the traditional login
> 4. Updating the token on the file, here I'm replacing the line on the file so at start make sure you have at least a dummy line on the host_var like `fortios_access_token: not_valid`. Once this task runs it will update with a valid token, make sure to double check the path too, it should make sense to whatever folder structure you're using.
> 
> And here's the same playbook but using the new auth api, availble on FOS 6.4.2+, there's some additional debug commands that you can remove if you want:
> 
> ```yaml
> - name: Login with credentials
>   ansible.builtin.uri:
>     # Authentication API available on FOS 6.4.2+
>     url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/api/v2/authentication
>     validate_certs: false
>     method: POST
>     body_format: json
>     body: {"username": "{{ ansible_user }}", "secretkey": "{{ ansible_password }}", "request_key": true}
>   register: login
> 
> - name: Configure API users
>   fortinet.fortios.fortios_system_api_user:
>     state: present
>     access_token: "{{ login.json.session_key }}"
>     system_api_user:
>       name: "{{ api_user_name }}"
>       accprofile: "{{ accprofile }}"
>       trusthost:
>         - id: "{{ trusthost_id }}"
>           ipv4_trusthost: "{{ ipv4_trusthost }}"
>       vdom:
>         - name: "{{ vdom }}"
> 
> - name: Generate API Key
>   ansible.builtin.uri:
>     url: https://{{ ansible_host }}:{{ ansible_httpapi_port }}/api/v2/monitor/system/api-user/generate-key
>     validate_certs: false
>     method: POST
>     body_format: json
>     body: {"api-user": "{{ api_user_name }}"}
>     headers:
>       Authorization: "Bearer {{ login.json.session_key }}"
>   register: r
> 
> - name: Set new token
>   set_fact:
>     fortios_access_token: "{{ r.json.results.access_token }}"
> 
> - name: Display new API key
>   debug:
>     var: fortios_access_token
> 
> - name: Update token at host_vars file
>   ansible.builtin.lineinfile:
>     path: "{{ ansible_env.PWD }}/../host_vars/{{ inventory_hostname }}.yml"
>     regexp: '^fortios_access_token:'
>     line: "fortios_access_token: {{ fortios_access_token }}"
> 
> - name: File updated
>   debug:
>     msg: "File {{ ansible_env.PWD }}/../host_vars/{{ inventory_hostname }}.yml updated"
> ```

Many thanks for your reply, i.e. the great example on how to do it with the uri method.

**Comment by chillancezen** [comment]
hi @carlos-segundo, @mbdraks ,

with our latest Galaxy collection which is recently released, we can also be able to generate access_token in various ways. 
please refer to the following docs:
https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#what-s-access-token
https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#where-to-keep-generated-access-token

feel free to reopen or create another issue if further assistance is needed. 
mark it `closed`.
thanks.
